feat: Add Snowflake destination plugin (#4965)
cloudquery/cloudquery: plugins/destination/snowflake/README.md

# CloudQuery Snowflake Destination Plugin

> The snowflake plugin helps you sync data to your Snowflake data warehouse. Note: This plugin is only for running the initial migrations and create the tables and schema. For the actual sync you need to use the file plugin and upload the data in csv/json to a storage bucket (S3/GCP/Azure Blob Storage) and use periodic `copy into` and/or snowpipe.

Author: File plugin is not out yet so we will have to update it once it is ready.

# CloudQuery Snowflake Destination Plugin

The snowflake plugin helps you sync data to your Snowflake data warehouse.

> There are two ways to sync data two Snowflake:

Reviewer: ```suggestion
There are two ways to sync data to Snowflake:
```

> 1. Direct (easy but not recommended for production or large data sets): This is the default mode of operation where CQ plugin will stream the results directly to the snowflake database. There is no additional setup needed apart from authentication to Snowflake.

Reviewer 2: ```suggestion
1. Direct (easy but not recommended for production or large data sets): This is the default mode of operation where CQ plugin will stream the results directly to the Snowflake database. There is no additional setup needed apart from authentication to Snowflake.
```

> 1. Direct (easy but not recommended for production or large data sets): This is the default mode of operation where CQ plugin will stream the results directly to the snowflake database, there is no additional setup needed apart from authentication to Snowflake.

Reviewer: ```suggestion
1. Direct (easy but not recommended for production or large data sets): This is the default mode of operation where CQ plugin will stream the results directly to the snowflake database. There is no additional setup needed apart from authentication to Snowflake.
```

> 2. Loading via CSV/Json from a remote Storage: This is a standard way of loading data into Snowflake, it is recommended for production and large data sets. This mode requires a remote storage (S3, GCS, Azure Blob Storage) and a Snowflake stage to be created. The CQ plugin will stream the results to the remote storage and then you can load those files via a cronjob or via SnowPipe. This way is still in the work and will be updated soon with a guide.

Reviewer: ```suggestion
2. Loading via CSV/Json from a remote Storage: This is a standard way of loading data into Snowflake, it is recommended for production and large data sets. This mode requires a remote storage (e.g. S3, GCS, Azure Blob Storage) and a Snowflake stage to be created. The CQ plugin will stream the results to the remote storage. You can then load those files via a cronjob or via SnowPipe. This method is still in the works and will be updated soon with a guide.
```

> 2. Loading via CSV/Json from a remote Storage: This is a standard way of loading data into Snowflake, it is recommended for production and large data sets. This mode requires a remote storage (e.g. S3, GCS, Azure Blob Storage) and a Snowflake stage to be created. The CQ plugin will stream the results to the remote storage. You can then load those files via a cronjob or via SnowPipe. This method is still in the works and will be updated soon with a guide.

Reviewer 2: ```suggestion
2. Loading via CSV/JSON from a remote storage: This is the standard way of loading data into Snowflake, it is recommended for production and large data sets. This mode requires a remote storage (e.g. S3, GCS, Azure Blob Storage) and a Snowflake stage to be created. The CQ plugin will stream the results to the remote storage. You can then load those files via a cronjob or via SnowPipe. This method is still in the works and will be updated soon with a guide.
```

BTW with snowflake you don't really need stages, `COPY INTO` can directly refer to an S3 bucket for instance (but you'd need access_key_id in the command which you might not want)

Author: I do think you still need stages at least per what I read recently even with remote storage you need to specify a stage first and mention that this stage related to a specific remote storage.

## Snowflake Spec

> This is the top level spec used by the Snowflake destination Plugin.

Reviewer 2: ```suggestion
This is the top level spec used by the Snowflake destination plugin.
```

- `dsn` (string) (required)

  Snowflake DSN

Example DSN:

```
	// user[:password]@account/database/schema[?param1=value1&paramN=valueN]
	// or
	// user[:password]@account/database[?param1=value1&paramN=valueN]
	// or
	// user[:password]@host:port/database/schema?account=user_account[?param1=value1&paramN=valueN]
	// or
	// host:port/database/schema?account=user_account[?param1=value1&paramN=valueN]

> ```

Reviewer: I think this needs to be indented one level deeper

Author: not sure there is any difference when markdown is rendered.

From Snowflake documentation:

> `account` - Name assigned to your Snowflake account. If you are not on us-west-2 or AWS deployment, append the region and platform to the end, e.g., <account>.<region> or <account>.<region>.<platform>.

Reviewer: Oh and this too

Author: Not sure as well if indent has any effect here

Reviewer: It changes whether it's in line with the `dsn` bullet or not (I guess it should be in line, because it's all still about the DSN)

Author: I see. ok indented.
